parallel_tempering.py

```diff
@@ -0,0 +1,254 @@
+
+###############################################################################
+# Functions below adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+###############################################################################
+
+
+"""
+Runs parallel tempering ***in serial*** for given inputs
+Called directly by controller/comparison files
+Arguments:
+    graph - a 2D numpy array where graph[A,B] is the distance from city A to
+            city B
+    function - a function that takes in values for graph and X_star and returns
+               an energy (E) value. ex: distance()
+    initial_Xs - initial ordered paths, one for each "system"
+    initial_temps - initial system temperatures, one for each "system", affect
+                    probability of update
+    iterr - number of iterations to perform for each system of parallel 
+            tempering algorithm
+    swap_function - a function that takes in values for X and nswaps and
+                    returns a new ordering of X. ex: changepath()
+    nswaps - number of swaps to perform when calculating updated path
+    nbefore - number of iterations prior to checking if systems should swap
+              values
+Returns:
+    -value of X with minimum associated value of E across all systems
+    -each system's history of E and X values
+"""
+def serial_parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                              iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are as expected
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
+
+    # Initialize stuff
+    nsystems = len(initial_temps)
+    Xs = list(initial_Xs)
+    Ts = initial_temps
+    prev_Es = [function(graph, Xs[i]) for i in range(nsystems)]
+    delta_Es = [0] * nsystems
+    history = [[] for i in xrange(nsystems)]
+    best_path = []
+    best_path_value = float('inf')
+
+    for i in xrange(nsystems):
+        history[i].append((prev_Es[i], initial_Xs[i]))
+
+    for step in range(iterr):
+        for i in range(nsystems):
+            # Run nbefore steps of simulated annealing
+            Xs[i], prev_Es[i], delta_Es[i], history[i] = anneal_once(graph,
+                                                        function, Xs[i], Ts[i], 
+                                                        prev_Es[i], history[i], 
+                                                        swap_function, nswaps)
+            # Store best path
+            if prev_Es[i] < best_path_value:
+                best_path = Xs[i]
+                best_path_value = prev_Es[i]
+
+        # Decide which chains, if any, to exchange
+        if step % nbefore == 0:
+            for i in range(nsystems - 1, 0, -1):
+                # Acceptance probability
+                # 0 = ln(1)                
+                A = np.exp(0, ((delta_Es[i] - delta_Es[i-1])/Ts[i]) +
```

Comment:
Should be
A = np.exp(min(0, ((delta_Es[i] - delta_Es[i-1])/Ts[i]) +
                              ((delta_Es[i-1] - delta_Es[i])/Ts[i-1])))